File: synapse_engine/lib/src/internal/jamo.dart
// Hangul jamo decomposition + Levenshtein distance.
// Port of engine/jamo.py — pure Dart, no external dependencies.

const int _hangulBase = 0xAC00;
const int _hangulEnd = 0xD7A3;
const int _jungCount = 21;
const int _jongCount = 28;

const List<String> _cho = [
  'ㄱ','ㄲ','ㄴ','ㄷ','ㄸ','ㄹ','ㅁ','ㅂ','ㅃ','ㅅ',
  'ㅆ','ㅇ','ㅈ','ㅉ','ㅊ','ㅋ','ㅌ','ㅍ','ㅎ',
];
const List<String> _jung = [
  'ㅏ','ㅐ','ㅑ','ㅒ','ㅓ','ㅔ','ㅕ','ㅖ','ㅗ','ㅘ',
  'ㅙ','ㅚ','ㅛ','ㅜ','ㅝ','ㅞ','ㅟ','ㅠ','ㅡ','ㅢ','ㅣ',
];
const List<String> _jong = [
  '','ㄱ','ㄲ','ㄳ','ㄴ','ㄵ','ㄶ','ㄷ','ㄹ','ㄺ',
  'ㄻ','ㄼ','ㄽ','ㄾ','ㄿ','ㅀ','ㅁ','ㅂ','ㅄ','ㅅ',
  'ㅆ','ㅇ','ㅈ','ㅊ','ㅋ','ㅌ','ㅍ','ㅎ',
];

/// Decomposes Hangul syllables into their cho/jung/jong jamo. Non-Hangul
/// characters are passed through unchanged.
String decompose(String text) {
  final buf = StringBuffer();
  for (final cp in text.runes) {
    if (cp >= _hangulBase && cp <= _hangulEnd) {
      final offset = cp - _hangulBase;
      final cho = offset ~/ (_jungCount * _jongCount);
      final jung = (offset % (_jungCount * _jongCount)) ~/ _jongCount;
      final jong = offset % _jongCount;
      buf.write(_cho[cho]);
      buf.write(_jung[jung]);
      if (jong > 0) buf.write(_jong[jong]);
    } else {
      buf.writeCharCode(cp);
    }
  }
  return buf.toString();
}

/// Standard iterative Levenshtein distance over code units.
int levenshtein(String s1, String s2) {
  if (s1.length < s2.length) return levenshtein(s2, s1);
  if (s2.isEmpty) return s1.length;

  var prev = List<int>.generate(s2.length + 1, (i) => i);
  for (var i = 0; i < s1.length; i++) {
    final curr = <int>[i + 1];
    for (var j = 0; j < s2.length; j++) {
      final cost = s1.codeUnitAt(i) == s2.codeUnitAt(j) ? 0 : 1;
      final ins = curr[j] + 1;
      final del = prev[j + 1] + 1;
      final sub = prev[j] + cost;
      var best = ins < del ? ins : del;
      if (sub < best) best = sub;
      curr.add(best);
    }
    prev = curr;
  }
  return prev.last;
}

/// Decomposes both strings to jamo, then compares with Levenshtein.
int jamoDistance(String a, String b) =>
    levenshtein(decompose(a), decompose(b));

/// Typo-candidate predicate. Both strings must be at least [minJamoLen]
/// jamo long after decomposition, the lengths must differ by at most
/// [maxDist], and the jamo distance must be ≤ [maxDist]. Identical strings
/// are never candidates.
bool isTypoCandidate(
  String a,
  String b, {
  int maxDist = 1,
  int minJamoLen = 6,
}) {
  if (a == b) return false;
  final ja = decompose(a);
  final jb = decompose(b);
  if (ja.length < minJamoLen || jb.length < minJamoLen) return false;
  if ((ja.length - jb.length).abs() > maxDist) return false;
  return levenshtein(ja, jb) <= maxDist;
}

File: archive/engine/jamo.py
"""한글 자모 분해 + Levenshtein 거리 유틸리티.

순수 파이썬, 외부 의존성 없음. 유니코드 수학으로 자모 분해.
"""

_HANGUL_BASE = 0xAC00
_HANGUL_END = 0xD7A3
_JUNG_COUNT = 21
_JONG_COUNT = 28

_CHO = list("ㄱㄲㄴㄷㄸㄹㅁㅂㅃㅅㅆㅇㅈㅉㅊㅋㅌㅍㅎ")
_JUNG = list("ㅏㅐㅑㅒㅓㅔㅕㅖㅗㅘㅙㅚㅛㅜㅝㅞㅟㅠㅡㅢㅣ")
_JONG = [""] + list("ㄱㄲㄳㄴㄵㄶㄷㄹㄺㄻㄼㄽㄾㄿㅀㅁㅂㅄㅅㅆㅇㅈㅊㅋㅌㅍㅎ")


def decompose(text: str) -> str:
    """문자열을 자모 단위로 분해. 비한글은 그대로 유지."""
    result: list[str] = []
    for ch in text:
        cp = ord(ch)
        if _HANGUL_BASE <= cp <= _HANGUL_END:
            offset = cp - _HANGUL_BASE
            cho = offset // (_JUNG_COUNT * _JONG_COUNT)
            jung = (offset % (_JUNG_COUNT * _JONG_COUNT)) // _JONG_COUNT
            jong = offset % _JONG_COUNT
            result.append(_CHO[cho])
            result.append(_JUNG[jung])
            if jong > 0:
                result.append(_JONG[jong])
        else:
            result.append(ch)
    return "".join(result)


def levenshtein(s1: str, s2: str) -> int:
    """두 문자열의 Levenshtein 편집거리."""
    if len(s1) < len(s2):
        return levenshtein(s2, s1)
    if len(s2) == 0:
        return len(s1)

    prev = list(range(len(s2) + 1))
    for i, c1 in enumerate(s1):
        curr = [i + 1]
        for j, c2 in enumerate(s2):
            cost = 0 if c1 == c2 else 1
            curr.append(min(curr[j] + 1, prev[j + 1] + 1, prev[j] + cost))
        prev = curr
    return prev[-1]


def jamo_distance(a: str, b: str) -> int:
    """두 문자열을 자모 분해 후 Levenshtein 거리 반환."""
    return levenshtein(decompose(a), decompose(b))


def is_typo_candidate(a: str, b: str, max_dist: int = 1, min_jamo_len: int = 6) -> bool:
    """자모 거리 기반 오타 후보 판정.

    조건: 양쪽 자모길이 >= min_jamo_len, 거리 <= max_dist, a != b.
    """
    if a == b:
        return False
    ja, jb = decompose(a), decompose(b)
    if len(ja) < min_jamo_len or len(jb) < min_jamo_len:
        return False
    if abs(len(ja) - len(jb)) > max_dist:
        return False
    return levenshtein(ja, jb) <= max_dist
